Finish implementation of compact index api draft #435

LukeIGS opened the pull request:
Closes: #435 

Saves the result of merging rubygems.org and local repo versions file to a file containing both indexes because this merging process takes a while and consumes a ton of memory to generate, so it should only be regenerated when necessary.

Also fix the concurrent implementation to function on windows.

Plus everything else @skaes implemented in the original draft.

skaes commented:
@LukeIGS I have more changes in my own upstream branch. Even caching the merged file is not fast enough for us. We now update the file on a cron job.

LukeIGS commented:
@skaes would it be possible to get a pr open for those in the near future?  It might be a good idea to have the ability to use the cronjob instead of having it merge them in flight as a flag, as with a small enough gem index it doesn't seem to be that slow.

LukeIGS commented:
> > @skaes would it be possible to get a pr open for those in the near future? It might be a good idea to have the ability to use the cronjob instead of having it merge them in flight as a flag, as with a small enough gem index it doesn't seem to be that slow.
> 
> The upstream index is always huge and this is what makes it slow. And not only slow, but it can bring down the server when you have some request parallelism. I would advise against it.

I can't say i've seen that at all.  I was thinking about implementing an incremental diff from the original file and the local index file to only update changed sections of the merged file though (a little like git does it)  would be significantly faster and less prone to a memory overflow. 

Thinking about it, the reason i likely don't see request parallelism kill it is because i have it deployed on k8s with an HPA on it.

ahorek commented on lib/geminabox/compact_indexer.rb:
```diff
@@ -0,0 +1,217 @@
+      say title unless progressbar_options
+
+      fork_type = {in_processes: n}
+      fork_type = {in_threads: n} if RUBY_PLATFORM == 'x64-mingw32'
```
```suggestion
      fork_type = {in_threads: n} unless Process.respond_to? :fork
```

ahorek commented on lib/geminabox/parallel_spec_reader.rb:
```diff
@@ -0,0 +1,44 @@
+        say title unless progressbar_options
+
+        fork_type = {in_processes: n}
+        fork_type = {in_threads: n} if RUBY_PLATFORM == 'x64-mingw32'
```
one more place :)
```suggestion
        fork_type = {in_threads: n} unless Process.respond_to? :fork
```

LukeIGS replied:
sml.

LukeIGS replied:
doesn't appear there are any more besides those two.

skaes commented:
> If i'm understanding your changes right, you basically just have a cronjob calling bin/geminabox reindex?

Not quite.

```
flock -n /tmp/geminabox.lock bundle exec geminabox update-cached-remote-specs
```

is the command.

LukeIGS commented:
Cool, i'm going to add that to the docs and add an an example k8s approach with this approach.

How often do you usually run it?

LukeIGS commented:
@marvinthepa
I currently run geminabox from the source branch of this pr (main).
```
git clone https://github.com/igs-opensource/geminabox
```
If you're using kubernetes i can look into packaging an example kubernetes deployment, otherwise, you can just 
`rackup --host 0.0.0.0` with a working directory of the root of the repo to run it.

Alternatively you can install it as a gem on your server from source by just calling `rake install` in the root directory

LukeIGS commented:
> @LukeIGS
> 
> > there's no way to stream the response to a tempfile
> 
> Sorry for hijacking this change for this tangent, but how about:
> 
> https://rubyapi.org/3.2/o/net/http#method-i-request_get
> 
> > When called with a block, passes an HTTPResponse object to the block. The body of the response will not have been read yet
> 
> The same is true for https://rubyapi.org/3.2/o/net/httpresponse#method-i-read_body:
> 
> > If a block is given, the body is passed to the block, and the body is provided in fragments, as it is read in from the socket.
> 
> so this should work:
> 
> ```ruby
> http.request_get('/index.html') do |response|
>   response.read_body do |segment|
>     print segment
>   end
> end
> ```
Ah, thanks for the pointer.  I actually noticed you can just dump the response directly to a file via open-uri as well, the issue right now is that most of the compact-index stuff is entirely encapsulated away from the api client, so i need to rewrite this somewhat to get either solution working.  At least not reading the entire body into memory would be a start though.

I did notice while poking around further that Geminabox seems to use HttpClient instead of Net::HTTP, but it seems like similar functionality is supported.  I'll probably poke around with it in the next few days when i have time.

nelsnelson commented:
@LukeIGS

> I opened an issue over at rubygems.org asking if we could get a streaming api implemented for version to reduce the memory footprint a bit ([rubygems/rubygems.org#3874](https://github.com/rubygems/rubygems.org/issues/3874))

[This](https://github.com/rubygems/rubygems.org/issues/3874#issuecomment-1686598967) update about client side bytes-pagination using etags looks exciting!  :D

LukeIGS commented:
yeah, etags are actually implemented already.  I didn't notice them right away, it simplified things quite a bit.  Atm i seem to have a mostly functional take on a pure in-file operation take on compact index merging, it's fairly quick, but more importantly uses very little memory.  I'm just updating the tests to reflect the new requirements i'm fulfilling and making sure i didn't break anything before i put it up.  Sadly i haven't had much time to devote to this lately so i may push this up to another branch in the IGS repo if someone else wants to take a poke at it.

nelsnelson commented:
> if someone else wants to take a poke at it.

I'd be happy to help.  Not sure what's left to be done.  Unit tests?  Automation for over-all functional testing?  I can try to contribute some of those as my free time permits.

LukeIGS commented:
Combination of a few things,  
1. satisfying code climate tests
2. Some of the specs defined in the compact index tests are unnecessary, such as the test for deduplication of gems in the version file where they get rewritten to be merged.  Rubygems.org uses append only logic and rewrites the entire thing every month or so, so just excluding any gems that the local server has with the same name as a remote gem is likely good enough.
3. Decision of actual specs that need to be fulfilled.  I don't fully trust myself to think of everything that should be tested and this is an open source project that has potentially larger impact.
4. Getting someone in the maintainers list to actually review and merge.

nelsnelson commented:
That does seems like a bit of work.  I'll take a look at the codeclimate stuff, and see what I can address.

LukeIGS commented:
https://github.com/igs-opensource/geminabox/tree/feature/compact-index-using-file-streams
Here's the current state, feel free to pr to this branch, i'll likely pr this into igsos:main after and then get the finished work into this pr.

natebird commented:
Any updates on this branch? Is there a workaround until this is merged?

LukeIGS commented:
running geminabox from https://github.com/igs-opensource/geminabox/tree/feature/compact-index-using-file-streams branch for the most part seems to work just fine for in the meantime as far as i can tell.  I've yet to run into any issues with it locally, though more testing would increase my confidence a bit.

ahorek commented:
hello @tnir is there any plan / hope to move this forward? it's likely bundler will remove support for old APIs sooner or later (rubygems.org already dropped support 5 months ago).
There's always room for further improvements, but is there anything important missing? Something we can fix / change / discuss? I understand it's a large feature, but it's just gathering dust for 3 years without any feedback from the maintainers.

as you can see, many users who want to use the new API are forced to install random forks which also isn't ideal... thanks for any reply :)

conathan commented:
Ran into some issues with a early version of the #435 patch (didn't investigate yet), so redid our setup to https://github.com/igs-opensource/geminabox/tree/feature/compact-index-using-file-streams.

Ran into a 500 error,

App 1784719 output: 2023-11-12 18:36:27 - TypeError - no implicit conversion of nil into String:
App 1784719 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-db3c501500a0/lib/geminabox/gem_versions_merge.rb:14:in `initialize'
App 1784719 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-db3c501500a0/lib/geminabox/gem_versions_merge.rb:14:in `open'
App 1784719 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-db3c501500a0/lib/geminabox/gem_versions_merge.rb:14:in `block in merge'
App 1784719 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-db3c501500a0/lib/geminabox/gem_versions_merge.rb:13:in `open'
App 1784719 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-db3c501500a0/lib/geminabox/gem_versions_merge.rb:13:in `merge'
App 1784719 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-db3c501500a0/lib/geminabox/compact_index_api.rb:37:in `versions'
App 1784719 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-db3c501500a0/lib/geminabox/server.rb:82:in `block in <class:Server>'
App 1784719 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1697:in `call'
App 1784719 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1697:in `block in compile!'
App 1784719 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1030:in `block (3 levels) in route!'
<pre>
lib/geminabox/gem_versions_merge.rb
    def self.merge(local_gem_list)
    ...
</pre>
* local_gem_list was nil

Also working off a new folder of /gems/rubygems.org, as wasn't sure if anything changed in the folder structure between that branch, and pull request #435 

config.ru
<pre>
require "rubygems"
require "geminabox"

Geminabox.data = "/gems/rubygems.org"
Geminabox.rubygems_proxy = true

Geminabox.allow_remote_failure = true
# No longer needed since patch 435
#Geminabox.rubygems_proxy_merge_strategy = :remote_gems_take_precedence_over_local_gems
run Geminabox
</pre>

This is a caching mirror, and no local gems.


---- what triggered it
AlmaLinux 9.2, with stock packages
bundler 2.2.33
ruby 3.0.4p208

Gemfile with:
gem "eventmachine", '1.2.7'
gem "rails", "7.1.1"
gem "sprockets-rails"
gem "bootsnap", require: false
gem "dalli"

LukeIGS commented:
That error basically happens when the local gem list either hasn't been populated or there are no gems on the server... i'll try to get a fix out for it.

LukeIGS commented:
@conathan Try the newest version i just pushed to that branch, it should now just generate an index if there isn't one already rather than requiring you to set one up yourself.  In reality this really just won't add much time on a fresh server.

conathan commented:
Ran into:

<pre>
App 1900072 output: 2023-11-13 17:59:17 - ArgumentError - wrong number of arguments (given 3, expected 2):
App 1900072 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:39:in `write_version_entries'
App 1900072 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:32:in `block (2 levels) in merge'
App 1900072 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:14:in `open'
App 1900072 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:14:in `block in merge'
App 1900072 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:13:in `open'
App 1900072 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:13:in `merge'
</pre>

lib/geminabox/gem_versions_merge.rb
Changed 
          write_version_entries(preamble, local_io, remote_versions_file)
to
          #write_version_entries(preamble, local_io, remote_versions_file) # Probably affected below line numbers by 1
          write_version_entries(local_io, remote_versions_file)
          
Then got:

<pre>
App 1900906 output: 2023-11-13 18:03:56 - EOFError - end of file reached:
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:49:in `readline'
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:49:in `block in write_version_entries'
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:92:in `block in try_load_cached_file'
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:94:in `open'
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:94:in `try_load_cached_file'
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:41:in `write_version_entries'
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:33:in `block (2 levels) in merge'
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:14:in `open'
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:14:in `block in merge'
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:13:in `open'
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:13:in `merge'
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/compact_index_api.rb:35:in `versions'
App 1900906 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/server.rb:82:in `block in <class:Server>'
</pre>

Wiped out the files I had in /gems/rubygems.org, as thought last exception could have caused it,  then got

<pre>
App 1901941 output: 2023-11-13 18:06:50 - ArgumentError - no time information in "at:":
App 1901941 output: 	/usr/share/ruby/time.rb:194:in `make_time'
App 1901941 output: 	/usr/share/ruby/time.rb:380:in `parse'
App 1901941 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:23:in `block (3 levels) in merge'
App 1901941 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:92:in `block in try_load_cached_file'
App 1901941 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:94:in `open'
App 1901941 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:94:in `try_load_cached_file'
App 1901941 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:20:in `block (2 levels) in merge'
App 1901941 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:14:in `open'
App 1901941 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:14:in `block in merge'
App 1901941 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:13:in `open'
App 1901941 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:13:in `merge'
App 1901941 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/compact_index_api.rb:35:in `versions'
App 1901941 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/server.rb:82:in `block in <class:Server>'

</pre>

lib/geminabox/gem_versions_merge.rb: self.write_version_entries

Changed
        preamble = "created at: #{Time.now}"
To
        preamble = "created_at: #{Time.now}"
(Wiped out files again for clean slate)
App 1903164 output: 	/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb:49:in `readline' (again)



<pre>
root@gems [/gems/rubygems.org/compact_index]# cat versions 
created_at: 2023-11-14T02:12:10.149+0000
---
root@gems [/gems/rubygems.org/compact_index]# cat merged_versions 
created_at: 2023-11-13 18:12:40 -0800---
</pre>
Noticed a missing newline after created_at:
</pre>
Changed in self.write_version_entries
        merged_versions_file.write("#{preamble}")
To
        merged_versions_file.write("#{preamble}\n")
(Suspect unrelated looking at the code).

Still same error, but wanted to post what I got so far.

conathan commented:
For the last issue I ran into,  local_gems_list is
 "created_at: 2023-11-14T02:20:24.633+0000\n---\n"

self.write_version_entries

does  a rewind, and 3 readline's (3rd being current_local_gem)
<pre>
irb(main):001:0> require 'stringio'
=> true
irb(main):002:0> s=StringIO.new "created_at: 2023-11-14T02:20:24.633+0000\n---\n"
=> #<StringIO:0x00007fb60c2ebd90>
irb(main):003:0> s.readline
=> "created_at: 2023-11-14T02:20:24.633+0000\n"
irb(main):004:0> s.readline
=> "---\n"
irb(main):005:0> s.readline
(irb):5:in `readline': end of file reached (EOFError)
	from (irb):5:in `<main>'
	from /usr/share/gems/gems/irb-1.3.5/exe/irb:11:in `<top (required)>'
	from /usr/bin/irb:23:in `load'
	from /usr/bin/irb:23:in `<main>'
</pre>

Not familiar enough with the codebase to suggest a fix for this,  but did:
<pre>
        current_local_gem = local_gem_io.readline rescue nil
        return nil if current_local_gem.nil?
</pre>
In our own environment  (and got what appears to be a successful bundle install going).

conathan commented:
Putting together everything as a diff.

<pre>
diff --git a/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b.orig/lib/geminabox/gem_versions_merge.rb b/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb
index 7d57901..c31c973 100644
--- a/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b.orig/lib/geminabox/gem_versions_merge.rb
+++ b/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb
@@ -29,7 +29,7 @@ module Geminabox
             end
           end
 
-          write_version_entries(preamble, local_io, remote_versions_file)
+          write_version_entries(local_io, remote_versions_file)
           File.read(merged_versions_path)
         end
       end
@@ -41,11 +41,12 @@ module Geminabox
         local_gem_list = local_gem_io.readlines.map { |it| it.split[0] }
         local_gem_io.rewind
         local_gem_io.readline # advance past preamble since we're writing our own merged versions file
-        preamble = "created at: #{Time.now}"
+        preamble = "created_at: #{Time.now}"
         merged_versions_file.flock(File::LOCK_EX)
-        merged_versions_file.write("#{preamble}")
+        merged_versions_file.write("#{preamble}\n")
         merged_versions_file.write(local_gem_io.readline) # get the ---
-        current_local_gem = local_gem_io.readline
+        current_local_gem = local_gem_io.readline rescue nil
+        return nil if current_local_gem.nil?
         remote_version_io.each_line do |remote_version_line|
           next if remote_version_line.eql?("---\n")
           remote_gem_name = remote_version_line.split[0]
</pre>

LukeIGS commented:
Just so i'm sure i understand this, the context here is that you're performing a bundle install with no local gems?  Just trying to figure out what process you're using so i can replicate it.

LukeIGS commented:
> Putting together everything as a diff.
> 
> diff --git a/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b.orig/lib/geminabox/gem_versions_merge.rb b/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb
> index 7d57901..c31c973 100644
> --- a/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b.orig/lib/geminabox/gem_versions_merge.rb
> +++ b/home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-b4e8a262200b/lib/geminabox/gem_versions_merge.rb
> @@ -29,7 +29,7 @@ module Geminabox
>              end
>            end
>  
> -          write_version_entries(preamble, local_io, remote_versions_file)
> +          write_version_entries(local_io, remote_versions_file)
>            File.read(merged_versions_path)
>          end
>        end
> @@ -41,11 +41,12 @@ module Geminabox
>          local_gem_list = local_gem_io.readlines.map { |it| it.split[0] }
>          local_gem_io.rewind
>          local_gem_io.readline # advance past preamble since we're writing our own merged versions file
> -        preamble = "created at: #{Time.now}"
> +        preamble = "created_at: #{Time.now}"
>          merged_versions_file.flock(File::LOCK_EX)
> -        merged_versions_file.write("#{preamble}")
> +        merged_versions_file.write("#{preamble}\n")
>          merged_versions_file.write(local_gem_io.readline) # get the ---
> -        current_local_gem = local_gem_io.readline
> +        current_local_gem = local_gem_io.readline rescue nil
> +        return nil if current_local_gem.nil?
>          remote_version_io.each_line do |remote_version_line|
>            next if remote_version_line.eql?("---\n")
>            remote_gem_name = remote_version_line.split[0]

okay, further updates are in the branch, one of the errors was caused by me forgetting to put an underscore between `created`and `at` in the merged file, so when it split it would try to parse `at:` to a time.

conathan commented:
That is correct.  Speaking of,  your latest changes fix all my issues.  Thumbsup.

conathan commented:
This is still not merged, (mostly replying for github actions)

conathan commented:
@LukeIGS Not sure where better to report this (then a closed ticket I don't think I can reopen), but was having troubles with some bundle installs (when ran in parallel via capistrano) saying things like ffi not found in any source.

And found this coresponding with it on geminabox
<pre>
App 373483 output: 2024-06-10 17:47:35 - ArgumentError - mon out of range:
App 373483 output:  /usr/share/ruby/time.rb:266:in `local'
App 373483 output:  /usr/share/ruby/time.rb:266:in `make_time'
App 373483 output:  /usr/share/ruby/time.rb:380:in `parse'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:23:in `block (3 levels) in merge'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:91:in `block in try_load_cached_file'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:93:in `open'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:93:in `try_load_cached_file'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:20:in `block (2 levels) in merge'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:14:in `open'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:14:in `block in merge'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:13:in `open'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:13:in `merge'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/compact_index_api.rb:35:in `versions'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/server.rb:82:in `block in <class:Server>'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1697:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1697:in `block in compile!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1030:in `block (3 levels) in route!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1049:in `route_eval'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1030:in `block (2 levels) in route!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1078:in `block in process_route'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1076:in `catch'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1076:in `process_route'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1028:in `block in route!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1025:in `each'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1025:in `route!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1147:in `block in dispatch!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1119:in `block in invoke'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1119:in `catch'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1119:in `invoke'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1142:in `dispatch!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:956:in `block in call!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1119:in `block in invoke'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1119:in `catch'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1119:in `invoke'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:956:in `call!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:945:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-protection-2.2.4/lib/rack/protection/xss_header.rb:18:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-protection-2.2.4/lib/rack/protection/path_traversal.rb:16:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-protection-2.2.4/lib/rack/protection/json_csrf.rb:26:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-protection-2.2.4/lib/rack/protection/base.rb:50:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-protection-2.2.4/lib/rack/protection/base.rb:50:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-protection-2.2.4/lib/rack/protection/frame_options.rb:31:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-2.2.8/lib/rack/null_logger.rb:11:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-2.2.8/lib/rack/head.rb:12:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-2.2.8/lib/rack/method_override.rb:24:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:218:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:2004:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1564:in `block in call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1780:in `synchronize'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1564:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox.rb:85:in `call'
App 373483 output:  /usr/share/ruby/vendor_ruby/phusion_passenger/rack/thread_handler_extension.rb:107:in `process_request'
App 373483 output:  /usr/share/ruby/vendor_ruby/phusion_passenger/request_handler/thread_handler.rb:149:in `accept_and_process_next_request'
App 373483 output:  /usr/share/ruby/vendor_ruby/phusion_passenger/request_handler/thread_handler.rb:110:in `main_loop'
App 373483 output:  /usr/share/ruby/vendor_ruby/phusion_passenger/request_handler.rb:419:in `block (3 levels) in start_threads'
App 373483 output:  /usr/share/ruby/vendor_ruby/phusion_passenger/utils.rb:113:in `block in create_thread_and_abort_on_exception'
</pre>

I printed out the results  of merged_versions_preamble from lib/geminabox/gem_versions_merge.rb,  I believe it is expected to be:
Created At: Jan 1 2024 (not exact)

But when it failed I was getting lines like
LanGrove 0.0.1,0.0.2 d37557b5d8ab1b6f85301f50d2618c59



I theorize when there are multiple requests at the same time, it can be reading the file while it is in the process of being created, so tried the following patch
<pre>
diff -Naurp geminabox-1c2fc3f28da5.orig/lib/geminabox/gem_versions_merge.rb geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb
--- geminabox-1c2fc3f28da5.orig/lib/geminabox/gem_versions_merge.rb     2024-06-11 10:14:34.312054817 -0700
+++ geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb  2024-06-11 10:20:24.289371650 -0700
@@ -19,6 +19,10 @@ module Geminabox
           preamble = (local_time > remote_time ? local_preamble : remote_preamble)
           try_load_cached_file do |merged_versions_file|
             unless merged_versions_file.eof?
+              #  - Multiple requests in parallel, first line is not created_at:
+              # This seems to resolve it
+              merged_versions_file.flock(File::LOCK_EX)
+              #  - End
               merged_versions_preamble = merged_versions_file.readline
               merged_version_time = Time.parse(merged_versions_preamble.split[1])
               preamble_time = Time.parse(preamble.split[1])
</pre>

And no longer able to reproduce the problem since (Appears to solve it, but I don't know the codebase well enough to say if it's the right solution)

conathan commented:
Also adding to this, we have been using @LukeIGS 's branch since mid November, 2023 with this being the only issue we encountered so far (Was working fine when used from a single server, but doing deploy's with bundle install's in parallel caused the issue reported in the previous comment).  Seems to be working well

LukeIGS commented:
I suspect you're right.  My gemstore is located on an NTFS storage device which generally tries to handle write collisions at an fs level.  If you want to open a pr against my fork I'll look into getting it merged in.

Month out of range implies that it tried to pull a nil date out of the file at some point which actually could happen if one thread read the file at the exact time another started to write it.

________________________________
From: Nathan Coulson ***@***.***>
Sent: Tuesday, June 11, 2024 3:00:14 PM
To: geminabox/geminabox ***@***.***>
Cc: Luke Ridge ***@***.***>; Mention ***@***.***>
Subject: Re: [geminabox/geminabox] Finish implementation of compact index api draft #435 (PR #547)

External Sender


@LukeIGS<https://github.com/LukeIGS> Not sure where better to report this (then a closed ticket I don't think I can reopen), but was having troubles with some bundle installs (when ran in parallel via capistrano) saying things like ffi not found in any source.

And found this coresponding with it on geminabox

App 373483 output: 2024-06-10 17:47:35 - ArgumentError - mon out of range:
App 373483 output:  /usr/share/ruby/time.rb:266:in `local'
App 373483 output:  /usr/share/ruby/time.rb:266:in `make_time'
App 373483 output:  /usr/share/ruby/time.rb:380:in `parse'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:23:in `block (3 levels) in merge'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:91:in `block in try_load_cached_file'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:93:in `open'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:93:in `try_load_cached_file'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:20:in `block (2 levels) in merge'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:14:in `open'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:14:in `block in merge'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:13:in `open'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb:13:in `merge'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/compact_index_api.rb:35:in `versions'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox/server.rb:82:in `block in '
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1697:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1697:in `block in compile!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1030:in `block (3 levels) in route!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1049:in `route_eval'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1030:in `block (2 levels) in route!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1078:in `block in process_route'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1076:in `catch'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1076:in `process_route'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1028:in `block in route!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1025:in `each'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1025:in `route!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1147:in `block in dispatch!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1119:in `block in invoke'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1119:in `catch'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1119:in `invoke'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1142:in `dispatch!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:956:in `block in call!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1119:in `block in invoke'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1119:in `catch'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1119:in `invoke'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:956:in `call!'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:945:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-protection-2.2.4/lib/rack/protection/xss_header.rb:18:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-protection-2.2.4/lib/rack/protection/path_traversal.rb:16:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-protection-2.2.4/lib/rack/protection/json_csrf.rb:26:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-protection-2.2.4/lib/rack/protection/base.rb:50:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-protection-2.2.4/lib/rack/protection/base.rb:50:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-protection-2.2.4/lib/rack/protection/frame_options.rb:31:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-2.2.8/lib/rack/null_logger.rb:11:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-2.2.8/lib/rack/head.rb:12:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/rack-2.2.8/lib/rack/method_override.rb:24:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:218:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:2004:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1564:in `block in call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1780:in `synchronize'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/gems/sinatra-2.2.4/lib/sinatra/base.rb:1564:in `call'
App 373483 output:  /home/webapps/geminabox-bn/shared/bundle/ruby/3.0.0/bundler/gems/geminabox-1c2fc3f28da5/lib/geminabox.rb:85:in `call'
App 373483 output:  /usr/share/ruby/vendor_ruby/phusion_passenger/rack/thread_handler_extension.rb:107:in `process_request'
App 373483 output:  /usr/share/ruby/vendor_ruby/phusion_passenger/request_handler/thread_handler.rb:149:in `accept_and_process_next_request'
App 373483 output:  /usr/share/ruby/vendor_ruby/phusion_passenger/request_handler/thread_handler.rb:110:in `main_loop'
App 373483 output:  /usr/share/ruby/vendor_ruby/phusion_passenger/request_handler.rb:419:in `block (3 levels) in start_threads'
App 373483 output:  /usr/share/ruby/vendor_ruby/phusion_passenger/utils.rb:113:in `block in create_thread_and_abort_on_exception'


I printed out the results of merged_versions_preamble from lib/geminabox/gem_versions_merge.rb, I believe it is expected to be:
Created At: Jan 1 2024 (not exact)

But when it failed I was getting lines like
LanGrove 0.0.1,0.0.2 d37557b5d8ab1b6f85301f50d2618c59

I theorize when there are multiple requests at the same time, it can be reading the file while it is in the process of being created, so tried the following patch

diff -Naurp geminabox-1c2fc3f28da5.orig/lib/geminabox/gem_versions_merge.rb geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb
--- geminabox-1c2fc3f28da5.orig/lib/geminabox/gem_versions_merge.rb     2024-06-11 10:14:34.312054817 -0700
+++ geminabox-1c2fc3f28da5/lib/geminabox/gem_versions_merge.rb  2024-06-11 10:20:24.289371650 -0700
@@ -19,6 +19,10 @@ module Geminabox
           preamble = (local_time > remote_time ? local_preamble : remote_preamble)
           try_load_cached_file do |merged_versions_file|
             unless merged_versions_file.eof?
+              #  - Multiple requests in parallel, first line is not created_at:
+              # This seems to resolve it
+              merged_versions_file.flock(File::LOCK_EX)
+              #  - End
               merged_versions_preamble = merged_versions_file.readline
               merged_version_time = Time.parse(merged_versions_preamble.split[1])
               preamble_time = Time.parse(preamble.split[1])


And no longer able to reproduce the problem since (Appears to solve it, but I don't know the codebase well enough to say if it's the right solution)

—
Reply to this email directly, view it on GitHub<https://github.com/geminabox/geminabox/pull/547#issuecomment-2161420568>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ARZYNIMR54U3V7FFWLJWSI3ZG5CLDAVCNFSM6AAAAAAYUHNS5KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNRRGQZDANJWHA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>



________________________________

Confidentiality Notice: The information contained in this email may be confidential and/or legally privileged. It has been sent for the sole use of the intended recipient(s). If you are not the intended recipient or authorized to receive information for the recipient, you are hereby notified that any review, use, disclosure, distribution, copying, printing, or action taken in reliance on the contents of this e-mail is strictly prohibited. If you have received this communication in error, please contact the sender by reply email and destroy all copies of the original message.

conathan commented:
@LukeIGS 
To give a simpler reproceable version of the error,

merged_versions_preamble = "LanGrove 0.0.1,0.0.2 d37557b5d8ab1b6f85301f50d2618c59"
Time.parse(merged_versions_preamble.split[1])

(or)
Time.parse("0.0.1,0.0.2")

<pre>
irb(main):001:0> merged_versions_preamble = "LanGrove 0.0.1,0.0.2 d37557b5d8ab1b6f85301f50d2618c59"
irb(main):002:0> Time.parse(merged_versions_preamble.split[1])
(irb):2:in `<main>': undefined method `parse' for Time:Class (NoMethodError)
	from /usr/share/gems/gems/irb-1.3.5/exe/irb:11:in `<top (required)>'
	from /usr/bin/irb:23:in `load'
	from /usr/bin/irb:23:in `<main>'
</pre>

More specifically, it's not grabbing the date, but one of the gems and trying to parse it's version as the date.

conathan commented:
@LukeIGS Will be my first time doing a PR, so I may not get to it immediately.  (in our case, I just patched locally, but thought useful to report upstream).

LukeIGS commented:
If you can't, I Should be able to look into it tomorrow.
________________________________
From: Nathan Coulson ***@***.***>
Sent: Tuesday, June 11, 2024 3:38:18 PM
To: geminabox/geminabox ***@***.***>
Cc: Luke Ridge ***@***.***>; Mention ***@***.***>
Subject: Re: [geminabox/geminabox] Finish implementation of compact index api draft #435 (PR #547)

External Sender


@LukeIGS<https://github.com/LukeIGS> Will be my first time doing a PR, so I may not get to it immediately. (in our case, I just patched locally, but thought useful to report upstream).

—
Reply to this email directly, view it on GitHub<https://github.com/geminabox/geminabox/pull/547#issuecomment-2161479028>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ARZYNIMNUONQMBZMRJ2GN33ZG5G2VAVCNFSM6AAAAAAYUHNS5KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNRRGQ3TSMBSHA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>


________________________________

Confidentiality Notice: The information contained in this email may be confidential and/or legally privileged. It has been sent for the sole use of the intended recipient(s). If you are not the intended recipient or authorized to receive information for the recipient, you are hereby notified that any review, use, disclosure, distribution, copying, printing, or action taken in reliance on the contents of this e-mail is strictly prohibited. If you have received this communication in error, please contact the sender by reply email and destroy all copies of the original message.

LukeIGS commented:
It appears i also, at some point made this patch locally and then forgot to upstream it.

conathan commented:
@LukeIGS I'll skip the PR then.

Hopefully someday this can get merged into geminabox, Seems like the github actions are a bit aggressive for closing these down (was a bit late to keep it open the last time)

wuarmin commented:
Hey @LukeIGS 
I started using geminabox today, cause I need to host some private gems (I host it in a docker-container). It also seems to work, but I get the html-output from sinatra that the API requests get a 404. 

```txt
<HTTP::Message::Headers:0x00007feeb1cbdfc0 @http_version="1.1", @body_size=0, @chunked=false, @request_method="GET", @request_uri=#<URI::HTTPS https://bundler.rubygems.org/api/v1/dependencies?
```

I don't know what the consequences of the error will be for me. I don't understand why the installation works anyway. What do you recommend? Should I use your branch, can I ignore the error?

Can you help me please?

Thanks

LukeIGS commented:
To explain what you're seeing, The dependency API was deprecated by rubygems.org.  It was replaced by a much more efficient solution called the compact index.  Bundler has an alternative solution it can fall back to called full index which takes a considerably longer time to resolve which is why the installation still works, but happens at a glacial pace.

You can ignore the error if you're okay with very slow gem installations at least for the forseeable future until rubygems.org deprecates the full index.

wuarmin commented:
Thanks @LukeIGS for the quick response. I appreciate it.

In the meantime I tried your branch, but I still get the error:

```ruby
source "https://rubygems.org"

gem "geminabox", git: "https://github.com/igs-opensource/geminabox.git", ref: "f44982e98b31d6826007db68ed3b209ed972c4ae"
gem "webrick", "~> 1.8"
```

```txt
 2024-09-11T14:21:51.562091679Z 172.16.6.17, 172.16.6.17 - - [11/Sep/2024:14:21:51 +0000] "GET /api/v1/dependencies?gems=activesupport%2Calba%2Cast%2Cawesome_print%2Caws-eventstream%2Caws-partitions%2Caws-sdk-core%2Caws-sdk-sqs%2Caws-sigv4%2Cbase64%2Cbetter_errors%2Cbigdecimal%2Cbinding_of_caller%2Ccoderay%2Cconcurrent-ruby%2Cconnection_pool%2Cdatabase_cleaner-core%2Cdatabase_cleaner-sequel%2Cdebug_inspector%2Cdiff-lcs%2Cdocile%2Cdotenv%2Cdrb%2Cdry-auto_inject%2Cdry-cli%2Cdry-configurable%2Cdry-core%2Cdry-events%2Cdry-files%2Cdry-inflector%2Cdry-initializer%2Cdry-logger%2Cdry-logic%2Cdry-monads%2Cdry-monitor%2Cdry-schema%2Cdry-struct%2Cdry-system%2Cdry-transformer%2Cdry-types%2Cdry-validation%2Cerubi%2Cfaker%2Cffi%2Cfiber-storage%2Cformatador%2Cgraphql%2Cgraphql-client%2Cguard%2Cguard-compat HTTP/1.1" 500 199268 0.1870

2024-09-11T14:21:51.730764204Z 2024-09-11 14:21:51 - HTTPClient::BadResponseError - unexpected response: #<HTTP::Message::Headers:0x00007f96c63cef60 @http_version="1.1", @body_size=0, @chunked=false, @request_method="GET", @request_uri=#<URI::HTTPS https://bundler.rubygems.org/api/v1/dependencies?gems=activesupport,alba,ast,awesome_print,aws-eventstream,aws-partitions,aws-sdk-core,aws-sdk-sqs,aws-sigv4,base64,better_errors,bigdecimal,binding_of_caller,coderay,concurrent-ruby,connection_pool,database_cleaner-core,database_cleaner-sequel,debug_inspector,diff-lcs,docile,dotenv,drb,dry-auto_inject,dry-cli,dry-configurable,dry-core,dry-events,dry-files,dry-inflector,dry-initializer,dry-logger,dry-logic,dry-monads,dry-monitor,dry-schema,dry-struct,dry-system,dry-transformer,dry-types,dry-validation,erubi,faker,ffi,fiber-storage,formatador,graphql,graphql-client,guard,guard-compat>, @request_query=nil, @request_absolute_uri=nil, @status_code=404, @reason_phrase="Not Found", @body_type=nil, @body_charset=nil, @body_date=nil, @body_encoding=#<Encoding:UTF-8>, @is_request=false, @header_item=[["Connection", "keep-alive"], ["Content-Length", "132"], ["Content-Type", "application/octet-stream; charset=utf-8"], ["x-frame-options", "SAMEORIGIN"], ["x-xss-protection", "0"], ["x-content-type-options", "nosniff"], ["x-permitted-cross-domain-policies", "none"], ["referrer-policy", "strict-origin-when-cross-origin"], ["cross-origin-opener-policy", "same-origin"], ["cache-control", "max-age=30, public"], ["content-encoding", "gzip"], ["content-security-policy", "default-src 'self'; font-src 'self' https://fonts.gstatic.com; img-src 'self' https://secure.gaug.es https://gravatar.com https://www.gravatar.com https://secure.gravatar.com https://*.fastly-insights.com https://avatars.githubusercontent.com; object-src 'none'; script-src 'self' 'sha256-VCdT6ytUhZs84Q7/wv/lrmONG3gBw7A6/v5HLZ1KLfU=' https://secure.gaug.es https://www.fastly-insights.com 'nonce-f508dcef1fd1ec80762df9b26049904b'; style-src 'self' https://fonts.googleapis.com 'nonce-f508dcef1fd1ec80762df9b26049904b'; connect-src 'self' https://s3-us-west-2.amazonaws.com/rubygems-dumps/ https://*.fastly-insights.com https://fastly-insights.com https://api.github.com http://localhost:*; form-action 'self' https://github.com/login/oauth/authorize; frame-ancestors 'self'; base-uri 'self'; report-uri https://csp-report.browser-intake-datadoghq.com/api/v2/logs?dd-api-key=pub852fa3e2312391fafa5640b60784e660&dd-evp-origin=content-security-policy&ddsource=csp-report&ddtags=service%3Arubygems.org%2Cversion%3A287e3eb784a50e8a63a9c71a1fac14901f2e0ae8%2Cenv%3Aproduction%2Ctrace_id%3A4267022885683641355"], ["x-request-id", "3805ab4e-2510-444b-9dae-1bdff4499f8a"], ["x-runtime", "0.002784"], ["strict-transport-security", "max-age=31536000"], ["X-Backend", "F_Rails 52.43.94.242:443"], ["Accept-Ranges", "bytes"], ["Date", "Wed, 11 Sep 2024 14:21:51 GMT"], ["Via", "1.1 varnish"], ["Age", "125"], ["X-Served-By", "cache-vie6346-VIE"], ["X-Cache", "HIT"], ["X-Cache-Hits", "2"], ["X-Timer", "S1726064512.725847,VS0,VE0"], ["Vary", "Accept-Encoding"], ["Server", "RubyGems.org"]], @dumped=false>:

2024-09-11T14:21:51.730893205Z 	/usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient.rb:1127:in `success_content'

2024-09-11T14:21:51.730909198Z 	/usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient.rb:655:in `get_content'

2024-09-11T14:21:51.730920931Z 	/usr/local/bundle/bundler/gems/geminabox-f44982e98b31/lib/geminabox/http_adapter/http_client_adapter.rb:14:in `get_content'

2024-09-11T14:21:51.730932244Z 	/usr/local/bundle/bundler/gems/geminabox-f44982e98b31/lib/geminabox/rubygems_dependency.rb:18:in `for'

2024-09-11T14:21:51.730942920Z 	/usr/local/bundle/bundler/gems/geminabox-f44982e98b31/lib/geminabox/server.rb:273:in `remote_gem_list'

2024-09-11T14:21:51.730953854Z 	/usr/local/bundle/bundler/gems/geminabox-f44982e98b31/lib/geminabox/server.rb:277:in `combined_gem_list'

2024-09-11T14:21:51.730964603Z 	/usr/local/bundle/bundler/gems/geminabox-f44982e98b31/lib/geminabox/server.rb:261:in `gem_list'

2024-09-11T14:21:51.730975150Z 	/usr/local/bundle/bundler/gems/geminabox-f44982e98b31/lib/geminabox/server.rb:68:in `block in <class:Server>'

2024-09-11T14:21:51.730986276Z 	/usr/local/bundle/gems/sinatra-2.2.4/lib/sinatra/base.rb:1697:in `call'
```
The error should no longer occur with your version, right?

LukeIGS commented:
That's biazarre, that stack looks like it's not using the new compact index flow at all.  I've never tried running it pulled as a packaged gem before, typically i'd recommend just pulling the source locally and running it via that.

wuarmin commented:
It must have something to do with the volume. When I comment it in, the error happens, when I run geminabox non-persistent everything works. Can you explain this?

```yml
services: 
  gem_server:
    image: private-registry.com/gem_server:current
    environment:
      DATA: /var/gem_server_data
    ports:
      - 9292:9292
    #volumes:
    #  - gem_server_data:/var/gem_server_data # If make the server persistent, the error happens
volumes:  
  gem_server_data:
    driver: local
    driver_opts:
      type: nfs
      o: addr=fs-puc1-nfs,nfsvers=4.1
      device: ":/vol/data"
```

LukeIGS commented:
What kind of infrastructure are you running on?  If it's something like AWS the host node may not allow local filesystem writes in which case you'd probably have to replace the volume driver with an s3 (it's been a long time since i've used docker swam so ymmv)

wuarmin commented:
It's just a debian hosts running docker engine. I'm using portainer to create docker swarm stacks. Filesystem writes should work, cause I saw files at the NFS share.

But maybe there are permission issues. If this is the case, can you explain the behavior?

LukeIGS commented:
I went ahead and enabled issues and discussions on the IGS mirror, since this is possibly an issue with our fork, lets move the discussion there.
